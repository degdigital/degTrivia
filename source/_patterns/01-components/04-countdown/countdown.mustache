{{# countdown }}
	<div class="countdown {{styleModifier}}">{{value}}</div>
{{/ countdown }}