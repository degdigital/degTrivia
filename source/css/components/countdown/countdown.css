.countdown {
	font-size: 2.5em;
	font-weight: normal;
}