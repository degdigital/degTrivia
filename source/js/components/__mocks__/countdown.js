function start() {
}

function stop() {
}

function __getInstance() {
	return instance;
}

const instance = {
	start,
	stop
};

//const countdown = opts => instance;
const countdown = jest.fn();
countdown.mockReturnValue(instance);
countdown.__getInstance = __getInstance;

export default countdown;