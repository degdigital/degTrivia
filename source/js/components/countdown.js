function convertFormat(settings, number, format = 'milliseconds') {
	if (number && settings.containerElement) {
		switch(format) {
			case 'milliseconds':
				return timer(number, settings.containerElement, settings.includeLabels);
			case 'seconds':
				return timer(number * 1000, settings.containerElement, settings.includeLabels);
			case 'minutes':
				return timer(number * 60 * 1000, settings.containerElement, settings.includeLabels);
				case 'hours':
				return timer(number * 60 * 60 * 1000, settings.containerElement, settings.includeLabels);
			case 'days':
				return timer(number * 60 * 60 * 24 * 1000, settings.containerElement, settings.includeLabels);             
		}
	}
}

function timer(milliseconds, containerElement, includeLabels) {
	const now = Date.now();
	const then = now + milliseconds;

	const countdown = setInterval(() => {
		const secondsLeft = Math.round((then - Date.now()) / 1000);

		if(secondsLeft < 0) {
			stopTimer(countdown);
			return;
		};

		displayTimeLeft(secondsLeft, containerElement, includeLabels);

	},1000);

	return countdown;
}

function renderUnit(num, label, includeLabels) {
	if (includeLabels) {
		return `
			<span class="countdown__time">${num}</span>
			<span class="countdown__unit">${label}${num === 1 || num === '01' ? '' : 's'}</span>
		`;
	}
	return num > 0 ? `<span class="countdown__time">${num}:</span>` : '';
	
}
	
function displayTimeLeft(seconds, containerElement, includeLabels) {
	containerElement.innerHTML = `
		${renderUnit(Math.floor(seconds / 86400), 'day', includeLabels)}
		${renderUnit(Math.floor((seconds % 86400) / 3600), 'hour', includeLabels)}
		${renderUnit(Math.floor((seconds % 86400) % 3600 / 60), 'minute', includeLabels)}
		${renderUnit(seconds % 60 < 10 ? `0${seconds % 60}` : seconds % 60, 'second', includeLabels)}
	`;
}

function stopTimer(countdown) {
	clearInterval(countdown);
}

export default function(opts) {
	const defaults = {
		containerElement: null,
		includeLabels: true
	}
	const settings = Object.assign({}, defaults, opts);
	let countdown;
	
	function start(num, format) {
		countdown = convertFormat(settings, num, format)
	}

	return {
		start,
		stop: () => stopTimer(countdown)
	}
};