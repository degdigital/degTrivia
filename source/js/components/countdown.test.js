import countdown from './countdown.js';

describe('countdown', () => {
    let countdownInst;

    beforeEach(() => {
        jest.useFakeTimers();
        document.body.innerHTML = '<div class="countdown-container"></div>';
        countdownInst = countdown();
    })

    afterEach(() => {
        setInterval.mockReset();
    })

    expect.extend({
        toBeEquivalent(received, argument) {
            const didPass = received.replace(/\s/g, '') === argument.replace(/\s/g, '');
            return {
                message: () => `expected ${received} to equal ${argument}`,
                pass: didPass
            }
        }
    })


    it('should not run if container el is not found', () => {
        document.body.innerHTML = '';
        countdownInst = countdown();
        countdownInst.start(1, 'seconds');
        expect(setInterval).toHaveBeenCalledTimes(0);
    });

    it('should not run if number param is undefined', () => {
        countdownInst.start();
        expect(setInterval).toHaveBeenCalledTimes(0);
    });

    it('should countdown each second', () => {
        countdownInst.start(1, 'seconds');
        expect(setInterval).toHaveBeenCalledTimes(1);
        jest.runAllTimers();
        expect(clearInterval).toHaveBeenCalledTimes(1);
    });

    it('should initialize coundown', () => {
        const initialOutput = `
        <div class="countdown-container">
            <span class="countdown__time">0</span>
            <span class="countdown__unit">days</span>
            <span class="countdown__time">0</span>
            <span class="countdown__unit">hours</span>
            <span class="countdown__time">0</span>
            <span class="countdown__unit">minutes</span>
            <span class="countdown__time">01</span>
            <span class="countdown__unit">second</span>
        </div>`;

        countdownInst.start(1, 'seconds');
        expect(setInterval).toHaveBeenCalledTimes(1);
        jest.advanceTimersByTime(1000);
        expect(document.body.innerHTML).toBeEquivalent(initialOutput);
    });

    it('should stop interval if called', () => {
        countdownInst.start(1, 'seconds');
        expect(setInterval).toHaveBeenCalledTimes(1);
        countdownInst.stop();
        expect(clearInterval).toHaveBeenCalledTimes(1);
    })
})