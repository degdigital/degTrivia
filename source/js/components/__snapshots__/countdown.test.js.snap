// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`countdown should initialize coundown 1`] = `
"<div class=\\"countdown-container\\">
			<span class=\\"countdown__time\\">0</span>
			<span class=\\"countdown__unit\\">days</span>
		
			<span class=\\"countdown__time\\">0</span>
			<span class=\\"countdown__unit\\">hours</span>
		
			<span class=\\"countdown__time\\">0</span>
			<span class=\\"countdown__unit\\">minutes</span>
		
			<span class=\\"countdown__time\\">01</span>
			<span class=\\"countdown__unit\\">second</span>
		</div>"
`;
